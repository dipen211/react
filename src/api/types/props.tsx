import { RouteComponentProps } from 'react-router-dom';
import {IFormState} from './types';

export interface CreateProps {
    RouteComponentProps: RouteComponentProps;
    IFormState: IFormState;
}