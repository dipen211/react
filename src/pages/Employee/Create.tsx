import React, { Component } from 'react';
import axios from 'axios';
import { RouteComponentProps } from 'react-router-dom';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import i18n from '../../translations/config';
import ValidationFormSchema from '../../validationSchema/ValidationFormSchema';
import {IFormState} from '../../api/types/types'

class Create extends Component<RouteComponentProps<any>, IFormState> {
    constructor(props: any) {
        super(props);
        this.state = {
            team_id: this.props.match.params.id,
            id: '',
            first_name: '',
            last_name: '',
            email: '',
            password: '',
            values: [],
            loading: false,
            submitSuccess: false,
        }
    }

    public render() {
        const { submitSuccess } = this.state;
        return (
            <div>
                <div className={"col-md-12 form-wrapper"}>
                    <h2> {i18n.t("cemployee")} </h2>
                    {!submitSuccess && (
                        <div className="alert alert-info" role="alert">
                            {i18n.t("cFillForm")}
                        </div>
                    )}
                    {submitSuccess && (
                        <div className="alert alert-info" role="alert">
                            {i18n.t("cSubmitSuccess")}
                        </div>
                    )}
                    <Formik
                        initialValues={this.state}
                        validationSchema={ValidationFormSchema}
                        onSubmit={async values => {
                            this.setState({ loading: true });
                            const formData = {
                                id: values.id,
                                team_id: this.state.team_id,
                                first_name: values.first_name,
                                last_name: values.last_name,
                                email: values.email,
                                password: values.password,
                            }
                            this.setState({ submitSuccess: true, values: [...this.state.values, formData], loading: false });
                            axios.post(`http://localhost:5000/employees`, formData).then(data => [
                                setTimeout(() => {
                                    this.props.history.push('/');
                                }, 1500)
                            ]);
                        }}>
                        {({ errors, touched, isSubmitting }) => {
                            return (
                                <Form>
                                    <div className="form-row">
                                        <div className="form-group col-6">
                                            <label>{i18n.t("id")}</label>
                                            <Field name="id" type="text" className={'form-control' + (errors.id && touched.id ? ' is-invalid' : '')} />
                                            <ErrorMessage name="id" component="div" className="invalid-feedback" />
                                        </div>
                                        <div className="form-group col-6">
                                            <label>{i18n.t("fname")}</label>
                                            <Field name="first_name" type="text" className={'form-control' + (errors.first_name && touched.first_name ? ' is-invalid' : '')} />
                                            <ErrorMessage name="first_name" component="div" className="invalid-feedback" />
                                        </div>
                                        <div className="form-group col-6">
                                            <label>{i18n.t("lname")}</label>
                                            <Field name="last_name" type="text" className={'form-control' + (errors.last_name && touched.last_name ? ' is-invalid' : '')} />
                                            <ErrorMessage name="last_name" component="div" className="invalid-feedback" />
                                        </div>
                                        <div className="form-group col-6">
                                            <label>{i18n.t("email")}</label>
                                            <Field name="email" type="text" className={'form-control' + (errors.email && touched.email ? ' is-invalid' : '')} />
                                            <ErrorMessage name="email" component="div" className="invalid-feedback" />
                                        </div>
                                    </div>
                                    <div className="form-group col-12">
                                        <label>{i18n.t("password")}</label>
                                        <Field name="password" type="password" className={'form-control' + (errors.password && touched.password ? ' is-invalid' : '')} />
                                        <ErrorMessage name="password" component="div" className="invalid-feedback" />
                                    </div>
                                    <div className="form-group">
                                        <button type="submit" disabled={isSubmitting} className="btn btn-primary">
                                            {isSubmitting && <span className="spinner-border spinner-border-sm mr-1"></span>}
                                            {i18n.t("save")}
                                        </button>
                                    </div>
                                </Form>
                            );
                        }}
                    </Formik>
                </div>
            </div>
        )
    }
}
export default (Create);
