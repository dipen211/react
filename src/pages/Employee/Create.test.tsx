import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { mountToJson } from 'enzyme-to-json';
import Create from './Create';
import { RouteComponentProps } from 'react-router-dom';
import { IFormState } from '../../api/types/types';
import { Formik } from 'formik';
Enzyme.configure({ adapter: new Adapter() });
const mockfn = jest.fn();

describe('Create', () => {
    let wrapper: any;
    const props: RouteComponentProps<{}> = {
        history: mockfn,
        location: mockfn,
        match: mockfn,
    }
    beforeEach(() => {
        wrapper = shallow(<Create {...props} />);
    });
    it('should match the snapshot', () => {
        expect(mountToJson(wrapper)).toMatchSnapshot();
    });
    it('should match the initial state', () => {
        expect(wrapper.instance().state.team_id).toBe(1);
        expect(wrapper.instance().state.id).toBe('');
        expect(wrapper.instance().state.first_name).toBe('');
        expect(wrapper.instance().state.last_name).toBe('');
        expect(wrapper.instance().state.email).toBe('');
        expect(wrapper.instance().state.password).toBe('');
        //expect(wrapper.instance().state.values).toBe([]);
        expect(wrapper.instance().state.loading).toBe(false);
        expect(wrapper.instance().state.submitSuccess).toBe(false);
    })

    it('should update firstname field on change', () => {
        wrapper.find('Field').at(0);
        wrapper.simulate('change');
        wrapper.setState({ first_name: "Dipen" })
        expect(wrapper.state('first_name')).toEqual('Dipen');
    });

    it('should update lastname field on change', () => {
        wrapper.find('Field').at(1);
        wrapper.simulate('change');
        wrapper.setState({ last_name: "Dipen" })
        expect(wrapper.state('last_name')).toEqual('Dipen');
    });

    it('should update email field on change', () => {
        wrapper.find('Field').at(2);
        wrapper.simulate('change');
        wrapper.setState({ email: "Dipen@gmail.com" })
        expect(wrapper.state('email')).toEqual('Dipen@gmail.com');
    });

    it('should update password field on change', () => {
        wrapper.find('Field').at(3);
        wrapper.simulate('change');
        wrapper.setState({ password: "Dipen" })
        expect(wrapper.state('password')).toEqual('Dipen');
    });

    it('should submit a valid form', () => {
        const state = {
            id: "1", team_id: "1",
            first_name: 'Dipen', last_name: 'Dipen',
            email: 'Dipen@gmail.com',
            password: "dipen",
        };
        const expectedArg = "id: 1, first_name: Dipen, last_name: Dipen, email: Dipen@gmail.com, password: dipen";
        window.alert = jest.fn();
        wrapper.setState(state);
        wrapper.find('Formik').simulate('submit', { preventDefault: () => { } });
        setTimeout(() => {
            wrapper.update();
        }, 0);
        expect(window.alert).toHaveBeenCalledWith(expectedArg);
    });

    it('should return error for invalid id', () => {
        const id = (props = { errors: {} }) =>
            wrapper
                .find(Formik)
                .renderProp('children')(props);

        const formWithInvalidIdErrors = id({
            errors: {
                id: 'id is invalid'
            },
            touched: { id: true },
            isSubmitting: false
        });

        expect(formWithInvalidIdErrors.html()).toEqual(/id is invalid/);
    });
    it('should return error for invalid first_name', () => {
        const first_name = (props = { errors: {} }) =>
            wrapper
                .find(Formik)
                .renderProp('children')(props);

        const formWithInvalidFirstNameErrors = first_name({
            errors: {
                first_name: 'first_name is invalid'
            },
            touched: { first_name: true },
            isSubmitting: false
        });

        expect(formWithInvalidFirstNameErrors.html()).toEqual("/first_name is invalid/");
    });

    it('should return error for invalid last_name', () => {
        const last_name = (props = { errors: {} }) =>
            wrapper
                .find(Formik)
                .renderProp('children')(props);

        const formWithInvalidLastNameErrors = last_name({
            errors: {
                last_name: 'last_name is invalid'
            },
            touched: { last_name: true },
            isSubmitting: false
        });

        expect(formWithInvalidLastNameErrors.html()).toEqual(/last_name is invalid/);
    });

    it('should return error for invalid email address', () => {
        const emailForm = (props = { errors: {} }) =>
            wrapper
                .find(Formik)
                .renderProp('children')(props);

        const formWithInvalidEmailErrors = emailForm({
            errors: {
                email: 'Email is invalid'
            },
            touched: { email: true },
            isSubmitting: false
        });

        expect(formWithInvalidEmailErrors.html()).toEqual(/Email is invalid/);
    });

    it('should return error for invalid password', () => {
        const passwordForm = (props = { errors: {} }) =>
            wrapper
                .find(Formik)
                .renderProp('children')(props);

        const formWithInvalidPasswordErrors = passwordForm({
            errors: {
                password: 'Password is invalid'
            },
            touched: { password: true },
            isSubmitting: false
        });

        expect(formWithInvalidPasswordErrors.html()).toEqual(/Password is invalid/);
    });
})