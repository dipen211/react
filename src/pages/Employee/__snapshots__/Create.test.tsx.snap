// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create should match the snapshot 1`] = `
<div>
  <div
    className="col-md-12 form-wrapper"
  >
    <h2>
       
      Create Employee
       
    </h2>
    <div
      className="alert alert-info"
      role="alert"
    >
      Fill the form to create Employee
    </div>
    <Formik
      initialValues={
        Object {
          "email": "",
          "first_name": "",
          "id": "",
          "last_name": "",
          "loading": false,
          "password": "",
          "submitSuccess": false,
          "team_id": 1,
          "values": Array [],
        }
      }
      onSubmit={[Function]}
      validationSchema={
        ObjectSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_nodes": Array [
            "password",
            "email",
            "last_name",
            "first_name",
            "id",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_sortErrors": [Function],
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "fields": Object {
            "email": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "email": false,
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "first_name": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "id": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "last_name": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
            "password": StringSchema {
              "_blacklist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "min": true,
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Set {},
                "refs": Map {},
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
              "type": "string",
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "object",
        }
      }
    >
      <Component />
    </Formik>
  </div>
</div>
`;
